src/cs50/sql.py

```diff
@@ -232,6 +258,38 @@ def execute(self, sql, *args, **kwargs):
         # Join tokens into statement
         statement = "".join([str(token) for token in tokens])
 
+        # Connect to database (for transactions' sake)
+        try:
+
+            # Infer whether Flask is installed
+            import flask
+
+            # Infer whether app is defined
+            assert flask.current_app
+
+            # If no connection for app's current request yet
+            if not hasattr(flask.g, "_connection"):
+
+                # Connect now
+                flask.g._connection = self._engine.connect()
+
+                # Disconnect later
+                @flask.current_app.teardown_appcontext
+                def shutdown_session(exception=None):
+                    flask.g._connection.close()
```

Comment:
@dmalan if the app context is popped after the request context is, shouldn't we pop `_connection` from `flask.g` before closing per https://flask.palletsprojects.com/en/1.1.x/appcontext/#storing-data or does that happen automatically by virtue of the fact that `g` is destroyed after the app context is popped?

Reply:
I think we're okay, per https://flask.palletsprojects.com/en/1.1.x/patterns/sqlalchemy/?

> Flask will automatically remove database sessions at the end of the request or when the application shuts down...

With that said, https://flask.palletsprojects.com/en/1.1.x/appcontext/ proposes popping:

> For example, you can manage a database connection using this pattern...

But former seems cleaner for now?